src/components/ContextMenu/index.js

```diff
@@ -0,0 +1,109 @@
+import React, { useState, useRef, useEffect, useLayoutEffect } from 'react';
+import styled from 'styled-components';
+import ContextMenuItem from './ContextMenuItem';
+
+function ContextMenu({ items, mousePos, parentRef, isToLeft, displayFocus }) {
+  const [isVisible, setIsVisible] = useState(false);
+  const [openToLeft, setOpenToLeft] = useState(false);
+  const [menuRef, setRefPos] = useState(null);
+  const [menuPos, setMenuPos] = useState({
+    top: '',
+    left: '',
+  });
+  const elementRef = useRef(null);
+
+  useLayoutEffect(() => {
+    if (displayFocus || parentRef)
+      setRefPos(elementRef.current.getBoundingClientRect());
+  }, []);
+
+  useEffect(() => {
+    initializeMenuPosition(mousePos, menuRef, parentRef);
+  }, [mousePos, menuRef]);
+
+  function initializeMenuPosition(mousePos, menuRef, parentRef) {
+    const windowWidth = window.innerWidth;
+    const windowHeight = window.innerHeight;
+    const newMenuPos = {
+      top: '',
+      left: '',
+    };
+
+    if (mousePos && menuRef) {
+      newMenuPos.top =
+        mousePos.y + menuRef.height > windowHeight
+          ? mousePos.y - menuRef.height
+          : mousePos.y;
+
+      newMenuPos.left =
+        mousePos.x + menuRef.width > windowWidth
+          ? mousePos.x - menuRef.width
+          : mousePos.x;
+
+      setMenuPos(prev => ({
+        ...prev,
+        top: newMenuPos.top,
+        left: newMenuPos.left,
+      }));
+      setIsVisible(true);
+      return;
+    }
+
+    if (menuRef && parentRef) {
+      // recalculate position for sub-menu
+      if (isToLeft) {
+        newMenuPos.left = -menuRef.width;
+        setOpenToLeft(true);
+      } else if (menuRef.x + menuRef.width + parentRef.width > windowWidth) {
+        newMenuPos.left = -menuRef.width;
+        setOpenToLeft(true);
+      } else {
+        newMenuPos.left = parentRef.width - 10;
+      }
+
+      //calculate bottom edge
+      newMenuPos.top =
+        menuRef.y + menuRef.height > windowHeight ? -menuRef.height + 20 : 0;
```

Comment:
💡 This `20` and the `10` above are "magic numbers" - I'm sure they have some meaning, but it's not obvious from the code. Consider naming them as constants above so it's easier to grok what they're about / when they should be updated.

Reply:
Fixed. I didn't add a variable for the number 0. It looks more confusing.